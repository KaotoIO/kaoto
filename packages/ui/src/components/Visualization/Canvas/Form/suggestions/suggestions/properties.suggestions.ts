import { SuggestionProvider } from '@kaoto/forms';

export const propertiesSuggestionProvider: SuggestionProvider = {
  id: 'properties-suggestion-provider',
  appliesTo: (_propName, schema) => schema.type === 'string',
  getSuggestions: async (word, _context) => {
    const normalizedWord = word !== '' ? word : 'foo';

    return [
      // https://camel.apache.org/manual/using-propertyplaceholder.html
      {
        value: `{{${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as a property reference`,
        group: word === '' ? 'Property references' : undefined,
      },
      {
        value: `{{${normalizedWord}:default}}`,
        description: `Use '${normalizedWord}' with a default value as a property reference`,
        group: 'Property references',
      },
      {
        value: `{{?${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as an optional property reference`,
        group: 'Property references',
      },
      {
        value: `{{!${normalizedWord}}}`,
        description: `Use '${normalizedWord}' as a negated property reference`,
        group: 'Property references',
      },
      {
        value: `RAW(${normalizedWord})`,
        description: `Use the RAW value of the '${normalizedWord}' property`,
        group: 'Property references',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_property_placeholder_functions
      {
        value: `{{env:${normalizedWord}}}`,
        description: `Use the '${normalizedWord}' OS environment variable`,
        group: 'Property functions',
      },
      {
        value: `{{sys:${normalizedWord}}}`,
        description: `Use the '${normalizedWord}' Java JVM system property`,
        group: 'Property functions',
      },
      {
        value: `{{service:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service, defined using the service naming idiom. Services are expected to be defined as OS ${normalizedWord.toUpperCase()}_SERVICE_HOST and ${normalizedWord.toUpperCase()}_SERVICE_PORT environment variables.`,
        group: 'Property functions',
      },
      {
        value: `{{service.name:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service hostname`,
        group: 'Property functions',
      },
      {
        value: `{{service.port:${normalizedWord.toUpperCase()}}}`,
        description: `Use the ${normalizedWord.toUpperCase()} service port`,
        group: 'Property functions',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_kubernetes_property_placeholder_functions
      {
        value: `{{configmap:${normalizedWord}}}`,
        description: `Use the ${normalizedWord} ConfigMap. It should be in the format 'configmap:configmap-name/configmap-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap:${normalizedWord}}}:defaultValue`,
        description: `Use the ${normalizedWord} ConfigMap with a default value`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap-binary:${normalizedWord}}}`,
        description: `Use the binary ${normalizedWord} ConfigMap`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret:${normalizedWord}}}`,
        description: `Use the ${normalizedWord} Secret. It should be in the format 'secret:secret-name/secret-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret-binary:${normalizedWord}}}`,
        description: `Use the binary ${normalizedWord} Secret.`,
        group: 'Kubernetes',
      },
    ];
  },
};
