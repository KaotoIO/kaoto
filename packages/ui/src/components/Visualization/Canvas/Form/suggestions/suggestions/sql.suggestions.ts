// Example of SQL syntax suggestion provider

import { SuggestionProvider } from '@kaoto/forms';
import { simpleLanguageSuggestionProvider } from './simple-language.suggestions';

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/sql.html
export const sqlSyntaxSuggestionProvider: SuggestionProvider = {
  ...simpleLanguageSuggestionProvider,
  id: 'sql-syntax-suggestion-provider',
  appliesTo: (propName, schema) => propName === '#.parameters.query' && schema.type === 'string',
  getSuggestions: async (word, _context) => {
    const suggestionWord = word !== '' ? word : 'foo';

    return [
      {
        value: `:#\${variable.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' variable`,
      },
      {
        value: `:#\${header.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' header`,
      },
    ];
  },
};
