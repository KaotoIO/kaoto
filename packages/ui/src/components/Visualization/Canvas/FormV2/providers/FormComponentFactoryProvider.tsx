import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
