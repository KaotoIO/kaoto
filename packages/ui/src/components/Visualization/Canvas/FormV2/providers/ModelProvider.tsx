import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
