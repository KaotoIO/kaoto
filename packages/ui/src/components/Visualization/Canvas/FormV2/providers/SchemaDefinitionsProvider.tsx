import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
