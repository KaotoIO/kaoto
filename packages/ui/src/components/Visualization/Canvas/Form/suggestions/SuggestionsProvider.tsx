import { useSuggestionRegistry } from '@kaoto/forms';
import { FunctionComponent, PropsWithChildren, useContext, useEffect } from 'react';
import { IMetadataApi, MetadataContext } from '../../../../../providers';
import { getPropertiesSuggestionProvider } from './suggestions/properties.suggestions';
import { getSimpleLanguageSuggestionProvider } from './suggestions/simple-language.suggestions';
import { sqlSyntaxSuggestionProvider } from './suggestions/sql.suggestions';

export const SuggestionRegistrar: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const suggestionRegistry = useSuggestionRegistry();
  const getSuggestions = useContext(MetadataContext)?.getSuggestions ?? GET_SUGGESTIONS_NOOP;

  useEffect(() => {
    const simpleLanguageSuggestionProvider = getSimpleLanguageSuggestionProvider(getSuggestions);
    const propertiesSuggestionProvider = getPropertiesSuggestionProvider(getSuggestions);

    suggestionRegistry?.registerProvider(simpleLanguageSuggestionProvider);
    suggestionRegistry?.registerProvider(propertiesSuggestionProvider);
    suggestionRegistry?.registerProvider(sqlSyntaxSuggestionProvider);

    return () => {
      suggestionRegistry?.unregisterProvider(simpleLanguageSuggestionProvider.id);
      suggestionRegistry?.unregisterProvider(propertiesSuggestionProvider.id);
      suggestionRegistry?.unregisterProvider(sqlSyntaxSuggestionProvider.id);
    };
  }, [getSuggestions, suggestionRegistry]);

  return <>{children}</>;
};

const GET_SUGGESTIONS_NOOP: IMetadataApi['getSuggestions'] = async () => {
  return [];
};
