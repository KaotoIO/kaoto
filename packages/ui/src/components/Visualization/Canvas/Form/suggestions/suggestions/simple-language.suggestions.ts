import { Suggestion, SuggestionProvider } from '@kaoto/forms';
import { CamelCatalogService, CatalogKind } from '../../../../../../models';
import { IMetadataApi } from '../../../../../../providers';

const SIMPLE_LANGUAGE_ACTIVATED_FIELDS = [
  '#.message', // This is a special case for the Log EIP
  'simple.expression',
];

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/simple-language.html
export const getSimpleLanguageSuggestionProvider = (metadata?: IMetadataApi['getSuggestions']): SuggestionProvider => {
  return {
    id: 'simple-language-suggestion-provider',
    appliesTo: (propName, schema) => SIMPLE_LANGUAGE_ACTIVATED_FIELDS.includes(propName) && schema.type === 'string',
    getSuggestions: async (word, context) => {
      const normalizedWord = word !== '' ? word : 'name';
      const normalizedWordLowercase = normalizedWord.toLowerCase();

      const simpleLangFunctions = CamelCatalogService.getComponent(CatalogKind.Function, 'simple') ?? {};
      const {
        body,
        ['variable.name']: variableName,
        ['header.name']: headerName,
        ['env.name']: envName,
        ...remainingFunctions
      } = simpleLangFunctions;

      const suggestions: Suggestion[] = [
        { value: `\${variable.${normalizedWord}}`, description: variableName?.description },
        { value: `\${header.${normalizedWord}}`, description: headerName?.description },
      ];

      const bodySuggestion: Suggestion = { value: '${body}', description: body?.description };
      if (word === '') {
        suggestions.unshift(bodySuggestion);
      } else {
        suggestions.push({ ...bodySuggestion, group: 'Simple Language' });
      }

      const environmentVariables = (await metadata?.('env', word, context)) ?? [];
      const environmentVariablesSuggestions = environmentVariables.map((item) => ({
        value: `\${env:${item.value}}`,
        description: `Use the '${item.value}' OS environment variable`,
        group: 'Simple Language: Environment variables',
      }));
      if (environmentVariablesSuggestions.length > 0) {
        suggestions.push(...environmentVariablesSuggestions);
      } else {
        suggestions.push({ value: `\${env.${normalizedWord}}`, description: envName?.description });
      }

      return suggestions.concat(
        Object.entries(remainingFunctions)
          .filter(([name]) => word === '' || name.toLowerCase().startsWith(normalizedWordLowercase))
          .map(([name, func]) => ({
            value: `\${${name}}`,
            description: func.description,
            group: 'Simple Language',
          })),
      );
    },
  };
};
