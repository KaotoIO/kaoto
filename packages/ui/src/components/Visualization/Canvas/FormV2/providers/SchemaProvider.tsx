import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
