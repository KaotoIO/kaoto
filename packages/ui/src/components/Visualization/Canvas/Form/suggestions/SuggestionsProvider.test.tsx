import { render } from '@testing-library/react';
import { ReactNode } from 'react';
import { IMetadataApi, MetadataContext } from '../../../../../providers';
import { SuggestionRegistrar } from './SuggestionsProvider';
import { sqlSyntaxSuggestionProvider } from './suggestions/sql.suggestions';
import { getPropertiesSuggestionProvider } from './suggestions/properties.suggestions';

// Mocks
const registerProvider = jest.fn();
const unregisterProvider = jest.fn();
const mockSuggestionRegistry = {
  registerProvider,
  unregisterProvider,
};

jest.mock('@kaoto/forms', () => ({
  useSuggestionRegistry: () => mockSuggestionRegistry,
}));
jest.mock('./suggestions/properties.suggestions', () => {
  const { getPropertiesSuggestionProvider } = jest.requireActual('./suggestions/properties.suggestions');
  const provider = getPropertiesSuggestionProvider(jest.fn());

  return {
    getPropertiesSuggestionProvider: jest.fn(() => provider),
  };
});
jest.mock('./suggestions/simple-language.suggestions', () => {
  const { getSimpleLanguageSuggestionProvider } = jest.requireActual('./suggestions/simple-language.suggestions');
  const provider = getSimpleLanguageSuggestionProvider(jest.fn());

  return {
    getSimpleLanguageSuggestionProvider: jest.fn(() => provider),
  };
});

describe('SuggestionRegistrar', () => {
  let mockMetadataApi: jest.Mocked<IMetadataApi>;

  beforeEach(() => {
    mockMetadataApi = {
      getMetadata: jest.fn(),
      setMetadata: jest.fn(),
      getResourceContent: jest.fn(),
      saveResourceContent: jest.fn(),
      deleteResource: jest.fn(),
      askUserForFileSelection: jest.fn(),
      getSuggestions: jest.fn(),
      shouldSaveSchema: false,
      onStepUpdated: jest.fn(),
    };
  });

  function renderWithMetadataProvider(children: ReactNode) {
    return render(
      <MetadataContext.Provider value={mockMetadataApi}>
        <SuggestionRegistrar>{children}</SuggestionRegistrar>
      </MetadataContext.Provider>,
    );
  }

  it('registers all providers on mount', () => {
    renderWithMetadataProvider(<div>Child</div>);

    const mockSimpleLanguageProvider = getPropertiesSuggestionProvider(mockMetadataApi.getSuggestions);
    const mockPropertiesProvider = getPropertiesSuggestionProvider(mockMetadataApi.getSuggestions);

    expect(registerProvider).toHaveBeenCalledWith(mockSimpleLanguageProvider);
    expect(registerProvider).toHaveBeenCalledWith(mockPropertiesProvider);
    expect(registerProvider).toHaveBeenCalledWith(sqlSyntaxSuggestionProvider);
  });

  it('unregisters all providers on unmount', () => {
    const { unmount } = renderWithMetadataProvider(<div>Child</div>);
    unmount();

    const mockSimpleLanguageProvider = getPropertiesSuggestionProvider(mockMetadataApi.getSuggestions);
    const mockPropertiesProvider = getPropertiesSuggestionProvider(mockMetadataApi.getSuggestions);

    expect(unregisterProvider).toHaveBeenCalledWith(mockSimpleLanguageProvider.id);
    expect(unregisterProvider).toHaveBeenCalledWith(mockPropertiesProvider.id);
    expect(unregisterProvider).toHaveBeenCalledWith(sqlSyntaxSuggestionProvider.id);
  });

  it('renders children', () => {
    const { container } = renderWithMetadataProvider(<span>Snapshot Child</span>);
    expect(container).toMatchSnapshot();
  });
});
