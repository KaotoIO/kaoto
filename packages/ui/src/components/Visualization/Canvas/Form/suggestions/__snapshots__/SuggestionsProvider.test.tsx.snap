// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionRegistrar renders children 1`] = `
<div>
  <span>
    Snapshot Child
  </span>
</div>
`;
